### PR TITLE
Fix issue 17019: `each` should be usable with parallel (and behave like foreach)

### PR DESCRIPTION
**Pull request by wilzbach**
The following test fails:

```d
unittest
{
    import std.algorithm.iteration : each, sum;
    import std.parallelism : parallel;
    import std.range : iota;

    auto logsIndex = new int[10];
    parallel(logsIndex).each!((i, ref e) => e += i);
    assert(logsIndex.sum == 10.iota.sum);
}
```

while it passes without parallel:

```d
unittest
{
    import std.algorithm.iteration : each, sum;
    import std.parallelism : parallel;
    import std.range : iota;

    auto logsIndex = new int[10];
    logsIndex.each!((i, ref e) => e += i);
    assert(logsIndex.sum == 10.iota.sum);
}
```

of course the foreach alternative works as well

```d
unittest
{
    import std.algorithm.iteration : sum;
    import std.parallelism : parallel;
    import std.range : iota;

    auto logsIndex = new int[10];
    foreach (i, ref e; parallel(logsIndex))
        e += i;
    assert(logsIndex.sum == 10.iota.sum);
}
```

(I also added more tests, which should hopefully help to cover more edge cases)

### CODE REVIEW

File: std/algorithm/iteration.d [hunk 1]
```diff
@@ -916,12 +923,20 @@ template each(alias pred = "a")
...
+            }
         }
-        else // if (isForeachBinaryIterable!Iterable)
+        else static if (isForeachBinaryIterable!Iterable)
         {
-            foreach (ref i, ref e; r)
+            foreach (ref a, ref b; r)
+                cast(void)binaryFun!BinaryArgs(a, b);
+        }
+        else // static if (isForeachUnaryWithIndexIterable!Iterable)
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: If you want, you can change the `static if` to a `static assert` and move it into the `else` block.

File: std/algorithm/iteration.d [hunk 2]
```diff
@@ -916,12 +923,20 @@ template each(alias pred = "a")
         debug(each) pragma(msg, "Using foreach for ", Iterable.stringof);
         static if (isForeachUnaryIterable!Iterable)
         {
+            debug(each) pragma(msg, "unary");
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: I guess this can be removed, now that you have debugged the issue ;)

File: std/algorithm/iteration.d [hunk 3]
```diff
@@ -916,12 +923,20 @@ template each(alias pred = "a")
...
             foreach (ref e; r)
+            {
                 cast(void)unaryFun!pred(e);
+            }
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: Nit/personal preference: curly braces are not needed for simple singe statement blocks.

File: std/algorithm/iteration.d [hunk 4]
```diff
@@ -1034,6 +1049,96 @@ template each(alias pred = "a")
     assert(res == [9, 12, 15]);
 }
 
+// #15357: `each` should behave similar to foreach
+/// each should work with index
+@safe unittest
+{
+    import std.range : iota, lockstep;
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: I think you also need to import `sum`, even though it's in same module, so the unittest can be run in the online documentation.

File: std/algorithm/iteration.d [hunk 5]
```diff
@@ -1034,6 +1049,96 @@ template each(alias pred = "a")
     assert(res == [9, 12, 15]);
 }
 
+// #15357: `each` should behave similar to foreach
+/// each should work with index
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: Since it's a ddoc-ed example, "`each` works with iterable objects which provide an index variable, along with each element", or sth. like that would be more appropriate.

File: std/algorithm/iteration.d [hunk 6]
```diff
@@ -1034,6 +1049,96 @@ template each(alias pred = "a")
...
+    auto arrA = [1, 2, 3, 4];
+    auto arrB = [5, 6, 7, 8];
+    lockstep(arrA, arrB).each!((ref a, ref b) {
+        a = 0;
+        b = 1;
+    });
+    assert(arrA.sum == 0);
+    assert(arrB.sum == 4);
+
+    /// three ref parameters
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: Nit: use either numbers or words on both places for consistency.

File: std/algorithm/iteration.d [hunk 7]
```diff
@@ -1034,6 +1049,96 @@ template each(alias pred = "a")
...
+{
+    import std.parallelism : parallel;
+    import std.range : iota;
+
+    // check behavior with parallel
+    auto logs = new int[10];
+    parallel(logs).each!((ref e) => e += 1);
+    assert(logs.sum == 10);
+
+    auto logsIndex = new int[10];
```
Comment (ZombineDev) [inline std/algorithm/iteration.d]: Nit: s/logs/arr/ (logs sounds confusing).

File: std/algorithm/iteration.d (hunk 1 above)
Comment (wilzbach) [inline std/algorithm/iteration.d]: Oh I opted for this because it was used like this before. I prefer `static assert` as well :)

File: std/algorithm/iteration.d (hunk 4 above)
Comment (wilzbach) [inline std/algorithm/iteration.d]: > even though it's in same module, so the unittest can be run in the online documentation.

Luckily only external dependencies (e.g. global imports), it's fairly trivial to a `import x.y.x` and we do so:

https://github.com/dlang/dlang.org/blob/master/js/run_examples.js#L29

The automatic tests extractor which is partially enabled for Phobos [does this as well](https://github.com/dlang/tools/blob/master/phobos_tests_extractor.d#L64) and will warn if someone tries to sneak public tests in that aren't runnable ;-)

Comment (ZombineDev) [inline std/algorithm/iteration.d]: Nice! LGTM then.

File: std/algorithm/iteration.d [hunk 8]
```diff
@@ -876,15 +876,22 @@ template each(alias pred = "a")
...
+            foreach (i, ref a; r)
+                cast(void)unaryFun!pred(i, a);
+        }));
+
     enum isForeachBinaryIterable(R) =
         is(typeof((R r) {
-            foreach (ref i, ref a; r)
-                cast(void)binaryFun!BinaryArgs(i, a);
+            foreach (ref a, ref b; r)
+                cast(void)binaryFun!BinaryArgs(a, b);
```
Comment (JackStouffer) [inline std/algorithm/iteration.d]: rationale for changing the parameter names?

Comment (dukc) [inline std/algorithm/iteration.d]: Strange, yes, but you can't be perfect. I think we should pull.

**Comment by wilzbach** [comment]
(rebased to latest `upstream/master`)

File: std/algorithm/iteration.d (hunk 8 above)
Comment (wilzbach) [inline std/algorithm/iteration.d]: To symbolize that it's a binary iteration with two elements - `i` is usually used as designated variable for the loop index.
The idea was to make it similar to `std.functional.binaryFun`.

**Comment by wilzbach** [comment]
That's weird. It seems for Windows this creates a cyclic dependency:

```
object.Error@src\rt\minfo.d(378): Cyclic dependency between module std.internal.math.biguintcore and std.parallelism
std.internal.math.biguintcore* ->
std.traits ->
std.algorithm.iteration ->
std.parallelism* ->
std.algorithm.iteration ->
std.bigint ->
std.internal.math.biguintcore*
```

CC @schveiguy any easy trick to workaround this?

**Comment by schveiguy** [comment]
@wilzbach I've run into that before. I think it's due to a difference in how the module info is output on non-Windows platforms. I never got around to figuring out why it happens. The cycle is definitely there, and it's not Windows-specific.

How you deal with the cycle is to get rid of the cycle, no easy tricks. If I were to approach it, I'd say try moving the CACHELIMIT variable and its initialization into its own file. That should break the cycle, since it only depends on druntime, so you can't have a cycle back to Phobos through core.cpuid.

**Comment by schveiguy** [comment]
> I've run into that before

FYI: https://github.com/dlang/druntime/pull/1602#issuecomment-230756312

**Comment by wilzbach** [comment]
Rebased to master. I have good hopes that this finally passes the testsuite as there was quite some work to get rid of the shared module constructors...

**Comment by wilzbach** [comment]
Yea, sometimes time does heal PRs :tada: 

image

**Comment by wilzbach** [comment]
> Yeah, sometimes time does heal PRs tada

Grr... and sometime time does break the PR again.
This time it's a segfault with -dip1000 (which is used to build this module since a few weeks):

Reduced ICE: https://issues.dlang.org/show_bug.cgi?id=18713

**Comment by wilzbach** [comment]
Now that we no longer use -dip1000, this passes again:

image

**Comment by wilzbach** [comment]
Can't be more green than now. Ready to move on?
